### DISCUSSION TITLE
apparmor denying access to USB from qdmr

### DISCUSSION THREAD

**Discussion by Altivo** [hmatuschek/qdmr#243] [General]
Linux Mint 20.3 (Una)
Removed apt restriction against installing snap by moving the file out of /etc/preferences.d
Then I could use "sudo apt install snapd" to get snap installed.
Followed by "sudo snap install qdmr" to install qdmr (which is apparently version 0.8.1)

[I guess I might need version 9.3 for this radio, but I'm not getting that far.]

Radio is TYT MD-UV390
Firmware version D019.005

USB cable provided with the new radio, part number 380DMR103-K

I did see the note buried in the manual that says the user must be in group dialout 
so ran "sudo adduser [myname] dialout" which executed without error.

However, when qdmr is launched with the radio plugged into USB, it still reports that it can't find the radio.
The syslog shows that the radio is recognized when plugged into USB, but apparmor is denying access from qdmr.

Syslog excerpt follows:

Aug 17 09:31:10 X200 kernel: [13925.684392] usb 1-5.2: new full-speed USB device
 number 10 using ehci-pci
Aug 17 09:31:10 X200 kernel: [13925.795183] usb 1-5.2: New USB device found, idV
endor=0483, idProduct=df11, bcdDevice= 2.00
Aug 17 09:31:10 X200 kernel: [13925.795189] usb 1-5.2: New USB device strings: M
fr=1, Product=2, SerialNumber=3
Aug 17 09:31:10 X200 kernel: [13925.795194] usb 1-5.2: Product: Digital Radio in USB mode
Aug 17 09:31:10 X200 kernel: [13925.795198] usb 1-5.2: Manufacturer: AnyRoad Technology
Aug 17 09:31:10 X200 kernel: [13925.795202] usb 1-5.2: SerialNumber: 00000000010C
...
Aug 17 09:31:28 X200 kernel: [13943.814479] audit: type=1400 audit(1660746688.265:784): apparmor="DENIED" operation="open" profile="snap.qdmr.qdmr" name="/sys/bus/usb/devices/" pid=3631 comm="qdmr" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
Aug 17 09:31:28 X200 kernel: [13943.815823] audit: type=1400 audit(1660746688.265:785): apparmor="DENIED" operation="open" profile="snap.qdmr.qdmr" name="/sys/bus/usb/devices/" pid=3631 comm="qdmr" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
Aug 17 09:31:28 X200 kernel: [13943.817917] audit: type=1400 audit(1660746688.265:786): apparmor="DENIED" operation="open" profile="snap.qdmr.qdmr" name="/run/udev/data/c4:28" pid=3631 comm="qdmr" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
Aug 17 09:31:28 X200 kernel: [13943.818379] audit: type=1400 audit(1660746688.265:787): apparmor="DENIED" operation="open" profile="snap.qdmr.qdmr" name="/run/udev/data/c4:79" pid=3631 comm="qdmr" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[many more repeats of similar messages]

It would seem that something is not correct in the apparmor configuration, but I know nothing about that functionality so I am at a loss to fix it.

Added information as of Aug 19:

Found a comment in the discussion area about a similar complaint from 2021. Instructions there said to tell snap to add access to raw USB data. Tried "sudo snap connect qdmr:raw-usb" as advised there, but still no success in getting a connection to the radio.

Here is the output from lsusb, in which the STM device in DFU mode does appear as device 8 on bus 1.

Bus 002 Device 004: ID 05ca:1880 Ricoh Co., Ltd USB2.0-FLASH Media    
Bus 002 Device 003: ID 05c6:9201 Qualcomm, Inc. Gobi Wireless Modem (QDL mode)
Bus 002 Device 002: ID 8086:1406 Intel Corp. WiMAX Connection 2400m
Bus 002 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 008 Device 001: ID 1d6b:0001 Linux Foundation 1.1 root hub
Bus 007 Device 001: ID 1d6b:0001 Linux Foundation 1.1 root hub
Bus 006 Device 001: ID 1d6b:0001 Linux Foundation 1.1 root hub
Bus 001 Device 006: ID 17ef:480c Lenovo Integrated Webcam
Bus 001 Device 008: ID 0483:df11 STMicroelectronics STM Device in DFU Mode        <==Noted here
Bus 001 Device 005: ID 17ef:1005 Lenovo 
Bus 001 Device 003: ID 0781:b6d0 SanDisk Corp. ExpCard Adpt
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 005 Device 001: ID 1d6b:0001 Linux Foundation 1.1 root hub
Bus 004 Device 002: ID 0a5c:2145 Broadcom Corp. BCM2045B (BDC-2.1) [Bluetooth Controller]
Bus 004 Device 001: ID 1d6b:0001 Linux Foundation 1.1 root hub
Bus 003 Device 002: ID 046d:c52f Logitech, Inc. Unifying Receiver
Bus 003 Device 001: ID 1d6b:0001 Linux Foundation 1.1 root hub

Also did install the CPS from the manufacturer as shipped with the radio on a CD, to Windows 7 on the same laptop. The manufacturer's CPS has the same problem, strangely enough, though Windows did see the device plugged in. A difference though: Windows complained that it could not find a suitable driver, even with an internet search.

**Comment by hmatuschek** [comment]
The group membership is only needed for the deb package to access radios implementing a serial-over-USB connection. 

For the USB access from the snap package, you have grand raw access to the application using
```
snap connect qdmr:raw-usb
```

**Reply by Altivo** [reply]
I did try that, as indicated above just before the lsusb output. The command granting raw access appears to run, but nothing changes.

I have now succeeded on Windows by manually installing a USB driver. The manufacturer's CPS is all but unusable because of the user interface and poor documentation.

On linux I am now successfully using editcp, from Dale Farnsworth at https://www.farnsworth.org/dale/codeplug/editcp/ 
so the hardware is in fact functional here. I suppose the problem is something in Linux Mint. The Mint maintainers state that they do not support snap.

**Reply by hmatuschek** [reply]
Maybe. Anyway, you could try to build qdmr from the sources or force to install the Ubuntu deb packages. With time, however, qdmr may pop up in the Mint repositories as it is now part of Debian and thus Ubuntu.

**Reply by Altivo** [reply]
Thanks for your work in building qdmr. I will continue to try with it. I have Debian also but it is on a 32 bit machine. Mint does have access to the Ubuntu repository, so that may do the job.
